text/0000-ubi-based-stacks.md

```diff
@@ -0,0 +1,181 @@
+appropriate runtimes baked in that would be selected by the Builder via Extensions. As with the previous option, 
+existing buildpacks would be checked to ensure they would use the baked in dependencies as a priority. 
+
+This approach leads to a very large Builder image, which is suboptimal from an end user perspective, plus, as all 
+runtimes are baked in, there would likely be frequent updates to the Builder image to incorporate fixes & security
+updates to each contained runtime, leading to the end-user requiring a fresh version of the builder image for almost 
+every update. A large image + frequent updates creates a less than ideal development experience.
+
+#### UBI without rpms
+This option would just take the existing Paketo buildpacks as-is and run them on a UBI base image. As with all the 
+options here, testing would be required for all contained buildpacks to ensure any system or shell interactions were 
+unaffected by the Ubuntu->UBI switch. 
+
+Although this approach would produce functional application images, the resulting images would not be using UBI 
+installed runtimes, and thus could be unable to qualify for full support in production. This is considered important 
+enough to end users that this option was discarded.
+
+### Chosen approach.
+The chosen approach of using an empty Builder image with Extension installed depdencies via RPM meets the goals of 
+having a lightweight development experience, while still being able to create supportable applications based on UBI 
+images. 
+
+## Implementation
+
+### Repositories
+
+The stacks team will create a new repo for UBI.. 
+
+* `ubi-base-stack`
+
+The buildpacks team will create two new repos for the initial two Extensions.
+
+* `ubi-java-extension`
+* `ubi-nodejs-extension`
+
+The *(unknown)* team will create the repo for the Builder image
+
+* `builder-ubi-base`
+
+## Prior Art
+
+No prior art for use of Extensions with Buildpacks to achieve this goal, beyond the prototypes created during 
+the CNCF Buildpack Extension development.
+
+## Unresolved Questions and Bikeshedding
+
+* Although the provisioning of dependencies via rpm/yum is a goal here, note it is not proposed to block the 
+use of dependencies downloaded via existing buildpacks. The rpm/yum behavior will be the default for the ubi stack. 
+
+* Note that multi-runtime applications (where say both nodejs and java form the final application) are not 
+covered by this stack. Although still achievable via end-user custom builders, it's not a goal for this incarnation. 
+
+* To allow for rebasing, and to work with the current Extension support, it is required to publish **multiple** 
+run images for a given builder image. This will likely require minor updates to the existing stack creation tooling. 
+
+* It may be possible in future builders to support run image customization via rpm/yum, at the cost of 
+losing rebasing support with the final application image. 
```

Comment:
These two bullets are my primary concerns.

1. The proliferation of run images and how much overhead that creates for the project. 

    @BarDweller Do you have any thoughts on ultimately how run images there will be? If not, is it a capped number or open-ended? 

    I'd also like to know what @paketo-buildpacks/stacks-maintainers think about the amount of overhead/toil this might create. Does it seem like it would be feasible to handle? Do we need more people-power to keep up? Do we perhaps need to look at adjustments to the tooling for efficiency purposes before or as part of accepting this RFC?

2. I'm concerned with the fact that upstream isn't done working on this feature and if we're rushing into something. 

    If we implement this now, we have to support what we're doing going forward for a reasonable amount of time. We can't just pivot and start doing something else, we'll be on the hook for support and publishing new run images for at least a year, if not longer. 

    So if in 6 months or a year, the buildpacks project finishes its work and run images support customization on-the-fly, what becomes of this implementation? Is it still the **best** implementation for our users?

Reply:
>That much said, I don't expect any great surprises from the way things are going upstream, if run-image extension becomes something that would offer customers value, it's not mutually exclusive within the same builder, there would just have to be env style config to select if the app image should use a precanned run image or allow the building of one during app build. (the latter coming with the cost of disabling rebasing).

I think you answered my next question too. I am wondering if you'd still suggest going ahead with the precanned run image even if, hypothetically, upstream had the dynamic run image extension support available today. It sounds like, yes, you still would. For users where rebasing is important.

It's more work to do both, but the idea of making that something a user could pick is interesting too. Precanned run images vs dynamic.

Reply:
Absolutely, the precanned enables rebasing, and I think that may be important enough to offer precanned for hand picked selection of runtimes.. for everything else, eventually, run-image customization will be 'enough' to start with, and if we put in the right sort of override, then an end user could force their own pre-canned run-image with the appropriate pre-installed customizations. 

Given where we're starting from though, it feels safe to have an initial plan to do precanned, and evolve it as upstream and requirements change =)